<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ cupcake</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">cupcake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/create-basis.html">Creating a basis using cupcake</a>
    </li>
    <li>
      <a href="../articles/project-basis.html">Projecting your data onto an exisiting basis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="LD.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SNP.manifest.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="add_ref_annotations.html">add_ref_annotations()</a></code> </p>
        </td>
        <td><p><code>add_ref_annotations</code> integrate GWAS summary data with support file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="align_alleles.html">align_alleles()</a></code> </p>
        </td>
        <td><p><code>align_alleles</code> compute variance weighted euclidean distance between two PC loading vectors a and b</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ambiguous.html">ambiguous()</a></code> </p>
        </td>
        <td><p><code>ambiguous</code> get ambiguous variants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="basis.trait.proj.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="beta.centers.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="comp.html">comp()</a></code> </p>
        </td>
        <td><p><code>comp</code> get complement for a set of variants</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_seb_proj_var.html">compute_seb_proj_var()</a></code> </p>
        </td>
        <td><p>analytically compute the variance of a projection given a standard error of beta for a given trait
<code>compute_seb_proj_var</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_seb_proj_var_sparse.html">compute_seb_proj_var_sparse()</a></code> </p>
        </td>
        <td><p>analytically compute the variance of a projection given a standard error of beta for a given trait for sparse basis
<code>compute_seb_proj_var_sparse</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="compute_shrinkage_metrics.html">compute_shrinkage_metrics()</a></code> </p>
        </td>
        <td><p>This function computes various shrinkage metrics
<code>compute_shrinkage_metrics</code> computes various shrinkage metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="control_prior_shape.html">control_prior_shape()</a></code> </p>
        </td>
        <td><p>This function computes shape parameters for beta distribution matching a binomial proportion B(n,f)/n
<code>control_prior_shape</code> shape parameters for beta distribution matching a binomial proportion B(n,f)/n</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cov_beta.html">cov_beta()</a></code> </p>
        </td>
        <td><p>covariance matrix of betas for an ld block
<code>cov_beta</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_basis.html">create_basis()</a></code> </p>
        </td>
        <td><p>This function creates a basis
<code>create_basis</code> creates a trait snp matrix that is suitable for basis creation and projection</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_ds_matrix.html">create_ds_matrix()</a></code> </p>
        </td>
        <td><p>This function creates a trait snp matrix
<code>create_ts_matrix</code> creates a trait snp matrix that is suitable for basis creation and projection</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="e_lor.html">e_lor()</a></code> </p>
        </td>
        <td><p>This function computes expected log odds ratio for a set of shape parameters
<code>e_lor</code> computes expected log odds ratio for a set of shape parameters</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="est_a1b1.html">est_a1b1()</a></code> </p>
        </td>
        <td><p>This function estimates shape parameters for a prior distribution of control allele frequencies
<code>est_a1b1</code> estimate shape parameters a1,b1</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="flip_allele.html">flip_allele()</a></code> </p>
        </td>
        <td><p><code>align_allele</code> given a data.table of pid,a1,a2 get a list of SNPs where alleles are flipped wrt to basis reference</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fopt.html">fopt()</a></code> </p>
        </td>
        <td><p>This function computes the optimal shape parameter a1 for a given set of constraints
shape parameters estimate the P(OR &gt; target.or).
<code>fopt</code> estimate the P(OR &gt; target.or) given shape parameters for case and control allele frequency</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_gwas_data.html">get_gwas_data()</a></code> </p>
        </td>
        <td><p><code>get_gwas_data</code> integrate GWAS summary data with support files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_seb.html">get_seb()</a></code> </p>
        </td>
        <td><p>An internal helper function to get obtain standard error of beta estimates from a GWAS.DT data.frame object
<code>get_seb</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logsum.html">logsum()</a></code> </p>
        </td>
        <td><p>helper function to sum logs without loss of precision
<code>logsum</code> sums logs without loss of precision</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lor_constraint.html">lor_constraint()</a></code> </p>
        </td>
        <td><p>This function computes a probability for a given configuration of beta distribution
shape parameters estimate the P(OR &gt; target.or).
<code>lor_constraint</code> estimate the P(OR &gt; target.or) given shape parameters for case and control allele frequency</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lor_f.html">lor_f()</a></code> </p>
        </td>
        <td><p>This function samples from the posterior distribution of a given control allele frequency, for different genotype configurations
<code>lor_f</code> sample from log(or) posterior distribution given an allele frequency.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lor_nf.html">lor_nf()</a></code> </p>
        </td>
        <td><p>This function calculates the posterior expectation of the log odds ratio for a number of case genotypes given control allele frequency</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lyons_egpa.html">lyons_egpa</a></code> </p>
        </td>
        <td><p>EGPA combined GWAS summary statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="maf_se_estimate.html">maf_se_estimate()</a></code> </p>
        </td>
        <td><p>Compute minor allele frequency shrinkage
<code>maf_se_estimate</code> computes a shrinkage metric for a given list of minor allele frequencies'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="maf_se_estimate_sample_size.html">maf_se_estimate_sample_size()</a></code> </p>
        </td>
        <td><p>Compute minor allele frequency shrinkage using sample size
<code>maf_se_estimate_sample_size</code> computes component of standard error of beta due to minor allele frequency</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mvs_perm.html">mvs_perm()</a></code> </p>
        </td>
        <td><p>Code to sample multivariate norm
<code>mvs_perm</code> sample from a multivariate normal distribution</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mvs_sigma.html">mvs_sigma()</a></code> </p>
        </td>
        <td><p>Code to compute sigma - genotype covariance matrix
<code>mvs_sigma</code> create a geneotype convariance matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="opt_a1b1.html">opt_a1b1()</a></code> </p>
        </td>
        <td><p>This function estimates shape parameters for a prior distribution of control allele frequencies
<code>opt_a1b1</code> estimate shape parameters a1,b1</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="p2z.html">p2z()</a></code> </p>
        </td>
        <td><p>convert p value  to a signed Z score
<code>p2z</code> p value to a signed Z score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="post_lor.html">post_lor()</a></code> </p>
        </td>
        <td><p>This function samples from the posterior distribution of log(OR) based on the observation of a case genotype
<code>post_lor</code> samples from the posterior distribution of log(OR) given genotype.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="project_basis.html">project_basis()</a></code> </p>
        </td>
        <td><p>This function projects an aligned trait onto the basis
<code>project_basis</code> projects an external trait into basis space</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="project_sparse.html">project_sparse()</a></code> </p>
        </td>
        <td><p>A streamlined function to project a trait onto a sparse basis
<code>project_sparse</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rot.pca.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="se_null.html">se_null()</a></code> </p>
        </td>
        <td><p>This function computes standard error under the null
<code>se_null</code> analytically compute standard error of \(\beta\) under \(\mathbb{E}(\beta) = 0\)
\(\sqrt{\frac{1}{2fN_0} +  \frac{1}{2N_{0}(1-f)} + \frac{1}{2fN_1} +  \frac{1}{2N_{1}(1-f)} }\)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shrinkage.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simulate_beta.html">simulate_beta()</a></code> </p>
        </td>
        <td><p>simulate betas for an ld block
<code>simulate_beta</code> use the multivariate normal to simulate realistic betas</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="simulate_study.html">simulate_study()</a></code> </p>
        </td>
        <td><p>simulate betas for a study
<code>simulate_study</code> use the multivariate normal to simulate realistic betas for a study</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="use.pca.html">burren_imd_13_traits</a></code> </p>
        </td>
        <td><p>Burren sparse basis files</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="vcf2snpmatrix.html">vcf2snpmatrix()</a></code> </p>
        </td>
        <td><p>convert a vcf file to snpMatrix object
<code>vcf2snpmatrix</code> convert a vcf file to snpMatrix object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wakefield_null_pp.html">wakefield_null_pp()</a></code> </p>
        </td>
        <td><p>compute reciprocal posterior probabilities using Wakefield's approximate Bayes Factors
<code>wakefield_null_pp</code> computes posterior probabilities for a given SNP to be NOT be causal for a given SNP under the assumption of a single causal variant.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="wakefield_pp.html">wakefield_pp()</a></code> </p>
        </td>
        <td><p>compute posterior probabilities using Wakefield's approximate Bayes Factors
<code>wakefield_pp</code> computes posterior probabilities for a given SNP to be causal for a given SNP under the assumption of a single causal variant.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ws_shrinkage.html">ws_shrinkage()</a></code> </p>
        </td>
        <td><p>This function computes an alternative to the Bayesian shrinkage method which can be too agressive.
<code>ws_shrinkage</code> computes a shrinkage based on a weighted sum (ws) of posteriors for each disease
this is then normalised by the total posterior for a given LD block</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="z2p.html">z2p()</a></code> </p>
        </td>
        <td><p>convert z to p value
<code>p2z</code> z to p value</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Olly Burren, Chris Wallace.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


